File: src/main/java/grokking/two_pointers/SortColors.java
package grokking.two_pointers;

import java.util.Arrays;

public class SortColors {
    public static int[] sortColors (int[] colors) {
        int start = 0,current=0;
        int end=colors.length-1;

        while (current <= end) {

            if (colors[current] == 0) {

                if (colors[start] != 0) {
                    int temp = colors[start];
                    colors[start] = colors[current];
                    colors[current] = temp;
                }

                current++;
                start++;
            }

            else if (colors[current] == 1) {
                current++;
            }

            else {
                if (colors[end] != 2) {
                    int temp = colors[current];
                    colors[current] = colors[end];
                    colors[end] = temp;
                }

                end--;
            }
        }

        return colors;
    }


    public static void main(String[] args) {
        System.out.println(Arrays.toString(sortColors(new int[]{0,1,0})));
    }
}

File: src/main/python/grokking/two_pointers/sort_colors.py
def sort_colors(colors: list[int]) -> list[int]:
    start = 0
    current = 0
    end = len(colors) - 1
    while current <= end:
        if colors[current] == 0:
            if colors[start] == 0:
                temp = colors[start]
                colors[start] = colors[current]
                colors[current] = temp
            current += 1
            start += 1

        elif colors[current] == 1:
            current += 1

        else:
            if colors[end] != 2:
                temp = colors[current]
                colors[current] = colors[end]
                colors[end] = temp

            end -= 1

    return colors
